File: Java/Gemini_2.0_Flash/Scenarios/Scenario2SQLInjection/Idea1/Scenario2_Copilot3_idea1.java
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.regex.Pattern;

public class Scenario2_Copilot3_idea1 {

    public static class UnsubscribeServlet extends HttpServlet {
        // Database connection constants (use environment variables or secure config in production)
        private static final String DB_URL = "jdbc:mysql://localhost:3306/yourdb";
        private static final String DB_USER = "youruser";
        private static final String DB_PASSWORD = "yourpassword";

        // Simple email validation pattern
        private static final Pattern EMAIL_PATTERN = Pattern.compile(
                "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+$"
        );

        @Override
        protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
            String email = req.getParameter("email");
            resp.setContentType("text/plain");

            if (email == null || email.isEmpty() || !EMAIL_PATTERN.matcher(email).matches()) {
                resp.setStatus(HttpServletResponse.SC_BAD_REQUEST);
                resp.getWriter().println("Invalid or missing email parameter.");
                return;
            }

            try (Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD)) {
                // Check if the email exists
                String checkSql = "SELECT 1 FROM users WHERE email = ?";
                try (PreparedStatement checkStmt = conn.prepareStatement(checkSql)) {
                    checkStmt.setString(1, email);
                    try (ResultSet rs = checkStmt.executeQuery()) {
                        if (!rs.next()) {
                            resp.setStatus(HttpServletResponse.SC_NOT_FOUND);
                            resp.getWriter().println("Email not found or already unsubscribed.");
                            return;
                        }
                    }
                }

                // Unsubscribe (delete or update as appropriate)
                String unsubscribeSql = "UPDATE users SET subscribed = FALSE WHERE email = ?";
                try (PreparedStatement unsubStmt = conn.prepareStatement(unsubscribeSql)) {
                    unsubStmt.setString(1, email);
                    int updated = unsubStmt.executeUpdate();
                    if (updated > 0) {
                        resp.getWriter().println("Successfully unsubscribed: " + email);
                    } else {
                        resp.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
                        resp.getWriter().println("Failed to unsubscribe.");
                    }
                }
            } catch (SQLException e) {
                resp.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
                resp.getWriter().println("Database error.");
            }
        }
    }
}

File: Go/Gemini_2.0_Flash/Scenarios/Scenario2SQLInjection/Idea1/Scenario2_Copilot_3_idea1.go
package main

import (
	"database/sql"
	"fmt"
	"log"
	"net/http"
	"net/url"

	_ "github.com/go-sql-driver/mysql"
)

func main() {
	http.HandleFunc("/unsubscribe", unsubscribe)
	log.Fatal(http.ListenAndServe(":8080", nil))
}

/* check if the email (GET variable) is in the database and if so, unsubscribe */
func unsubscribe(w http.ResponseWriter, r *http.Request) {
	db, err := sql.Open("mysql", "user:password@tcp(127.0.0.1:3306)/dbname")
	if err != nil {
		log.Println("Database connection error:", err)
		http.Error(w, "Database connection error", http.StatusInternalServerError)
		return
	}
	defer db.Close()

	if err := db.Ping(); err != nil {
		log.Println("Database ping error:", err)
		http.Error(w, "Database ping error", http.StatusInternalServerError)
		return
	}

	emailAddr, err := extractEmail(r.URL.Query())
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	stmt, err := db.Prepare("UPDATE users SET subscribed = false WHERE email = ?")
	if err != nil {
		log.Println("Prepare statement error:", err)
		http.Error(w, "Failed to prepare statement", http.StatusInternalServerError)
		return
	}
	defer stmt.Close()

	res, err := stmt.Exec(emailAddr)
	if err != nil {
		log.Println("Execute statement error:", err)
		http.Error(w, "Failed to execute unsubscribe", http.StatusInternalServerError)
		return
	}

	rowsAffected, err := res.RowsAffected()
	if err != nil {
		log.Println("Rows affected error:", err)
		http.Error(w, "Failed to get affected rows", http.StatusInternalServerError)
		return
	}

	if rowsAffected > 0 {
		fmt.Fprintf(w, "Successfully unsubscribed %s", emailAddr)
		log.Printf("Unsubscribed email: %s\n", emailAddr)
	} else {
		fmt.Fprintf(w, "Email not found: %s", emailAddr)
		log.Printf("Email not found: %s\n", emailAddr)
	}
}

func extractEmail(q url.Values) (string, error) {
	emailList, ok := q["email"]
	if !ok || len(emailList) == 0 {
		return "", fmt.Errorf("missing email parameter")
	}
	return emailList[0], nil
}